improver/utilities/field_texture.py

```diff
@@ -0,0 +1,202 @@
+        )
+
+        # Calculate the actual transitions for each grid cell of value 1 and
+        # store them in a cube.
+        actual_transitions = potential_transitions.copy(
+            data=self._calculate_transitions(cube.data)
+        )
+
+        # Sum the number of actual transitions within the neighbourhood.
+        actual_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            actual_transitions, radius=radius
+        )
+        actual_transitions.data = np.where(cube.data > 0, actual_transitions.data, 0)
+
+        # Calulate the ratio of actual to potential transitions
+        ratio = np.ones_like(actual_transitions.data)
+        np.divide(
+            actual_transitions.data,
+            potential_transitions.data,
+            out=ratio,
+            where=(cube.data > 0),
+        )
+
+        # Create a new cube to contain the resulting ratio data
+        ratio = create_new_diagnostic_cube(
+            "cell_edge_differences", 1, cube, MANDATORY_ATTRIBUTE_DEFAULTS, data=ratio,
+        )
+        return ratio
+
+    @staticmethod
+    def _calculate_transitions(data):
+        """
+        Identifies actual transitions present in a binary field. These transitions
+        are defined as the number of cells of value zero that directly neighbour
+        cells of value 1. The number of transitions is calculated for all cells
+        of value 1. The number of transitions for cells of value 0 is set to 0.
+
+        Args:
+            data (numpy.ndarray):
+                A NumPy array of the input cube for data manipulation.
+
+        Returns:
+            numpy.ndarray:
+                A NumPy array containing the transitions for ratio calculation.
+        """
+        padded_data = np.pad(data, 1, mode="edge")
+        diff_x = np.abs(np.diff(padded_data, axis=1))
+        diff_y = np.abs(np.diff(padded_data, axis=0))
+        cell_sum_x = diff_x[:, 0:-1] + diff_x[:, 1:]
+        cell_sum_y = diff_y[0:-1, :] + diff_y[1:, :]
+        cell_sum = cell_sum_x[1:-1, :] + cell_sum_y[:, 1:-1]
+        cell_sum = np.where(data > 0, cell_sum, 0)
+        return cell_sum
+
+    def process(self, cube):
+        """
+        Calculates a field of texture to use in differentiating solid and
+        more scattered features.
+
+        Args:
+            cube (iris.cube.Cube):
+                Input data in cube format containing the field for which the
+                texture is to be assessed.
+
+        Returns:
+            iris.cube.Cube:
+                A cube containing the mean of the thresholded ratios in cube
+                format.
+        """
+        ratios = iris.cube.CubeList()
+
+        try:
+            cslices = cube.slices_over("realization")
+        except CoordinateNotFoundError:
+            cslices = [cube]
+        for cslice in cslices:
+            ratios.append(self._calculate_ratio(cslice, self.nbhood_radius))
+        ratios = ratios.merge_cube()
```

Comment:
The metadata on the cube being created here are not correct.  What name does this cube have?  What units?  It's not the same as the input cube.

This plugin is creating a new diagnostic - the field texture.  It should be setting up a new cube with a distinct name and units (should be calculated from the units of the original diagnostic if need be), using `improver.metadata.utilities.create_new_diagnostic_cube`.

Ideally all of the private methods in this class would work in arrays.  Actually, given the way the neighbourhooding code is written, I'm not sure that's possible.  But if it's working in cubes, the methods here should not be copying cube slices from the inputs, but should be using the new cube creation utility as above.

Reply:
Note @cgsandford that ```create_new_diagnostic_cube``` is being used within ```_calculate_ratio```, such that the cube being appended to ratio is indeed a new cube with a new name, units, etc.

Reply:
Ah sorry, I missed that.  The note on suitable metadata stands, though.

Reply:
This thread got a bit confused so just clarifying.  The change that's required here is to give the output cube a suitable name (currently this code will output `probability_of_cell_edge_differences_above_threshold`, which may be accurate but isn't very useful!).  How this is best done depends on how the code is written.  So I think we need to wait until @bayliffe has decided on a suitable name for the output of this plugin, and then we can go through your code and see about the best place to do that.

Reply:
Done.